xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-index/src/main/java/org/xwiki/extension/index/internal/listener/ExtensionIndexCleanupListener.java

```diff
@@ -0,0 +1,136 @@
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrCoreInitializer;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrUtil;
+import org.xwiki.extension.repository.CoreExtensionRepository;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.ApplicationStartedEvent;
+import org.xwiki.observation.event.Event;
+import org.xwiki.search.solr.Solr;
+import org.xwiki.search.solr.SolrException;
+import org.xwiki.search.solr.SolrUtils;
+
+import static org.xwiki.extension.index.internal.ExtensionIndexSolrCoreInitializer.IS_INSTALLED_EXTENSION;
+import static org.xwiki.search.solr.AbstractSolrCoreInitializer.SOLR_FIELD_ID;
+
+/**
+ * Cleanup the extension index of entries related to artifacts that are not available anymore (e.g., upgrade or manual
+ * manipulation of the directories storing jar files).
+ *
+ * @version $Id$
+ * @since 15.9RC1
+ * @since 15.5.3
+ */
+@Component
+@Singleton
+@Named("ExtensionIndexCleanupListener")
+public class ExtensionIndexCleanupListener implements EventListener
+{
+    @Inject
+    private CoreExtensionRepository coreExtensionRepository;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private Solr solr;
+
+    @Inject
+    private SolrUtils solrUtils;
+
+    @Inject
+    private ExtensionIndexSolrUtil extensionIndexSolrUtil;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public String getName()
+    {
+        return "ExtensionIndexCleanupListener";
+    }
+
+    @Override
+    public List<Event> getEvents()
+    {
+        return List.of(new ApplicationStartedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        Thread thread = new Thread(this::proceed);
+        thread.setPriority(thread.getPriority() - 2);
+        thread.start();
+    }
+
+    private void proceed()
+    {
+        try {
+            SolrClient client = this.solr.getClient(ExtensionIndexSolrCoreInitializer.NAME);
+            SolrQuery solrQuery = new SolrQuery()
+                // Only return the id and the installed extensions fields as they are the only fields we are interested
+                // in.
+                .setFields(SOLR_FIELD_ID, IS_INSTALLED_EXTENSION);
+            int batchSize = 1000;
+            QueryResponse search = client.query(solrQuery.setRows(batchSize).setStart(0));
+
+            while (!search.getResults().isEmpty()) {
+                for (SolrDocument doc : search.getResults()) {
+                    String id = this.solrUtils.getId(doc);
+                    ExtensionId extensionId = this.extensionIndexSolrUtil.fromSolrId(id);
+                    boolean isCoreExtension = Objects.equals(doc.getFieldValue(IS_INSTALLED_EXTENSION), false);
```

Comment:
Are you sure this is enough to identify a core extension ? This index is full of extensions which are neither installed nor core extensions (its initial job was to index remote extension was can install).

Reply:
When false, I thinks it's accurate. I'm less sure when it's null.

Reply:
I feel like we should have a more explicit way to indicate what is a core extension, ideally.